lib/grp.gi

```diff
@@ -4291,86 +4299,125 @@ end );
-      fi;
-      typ:=fam!.defaultFinitelyGeneratedGroupType;
-    else
-      if not IsBound(fam!.defaultGroupType) then
-        fam!.defaultGroupType:=
-          NewType(fam,IsGroup and IsAttributeStoringRep
-                      and HasGeneratorsOfMagmaWithInverses);
+function( gens )
+local G,fam,typ,id,typename;
+
+  fam:=FamilyObj(gens);
+
+  typ:=IsGroup and IsAttributeStoringRep 
+    and HasIsEmpty and HasGeneratorsOfMagmaWithInverses;
+  typename:="defaultGroupType";
+
+  if IsFinite(gens) then
+    typ:=typ and IsFinitelyGeneratedGroup;
+    typename:="defaultFinitelyGeneratedGroupType";
+
+    if Length(gens)>0 and CanEasilyCompareElements(gens) then
+      id:=One(gens[1]);
+      if ForAny(gens,x->x<>id) then
+        typ:=typ and HasIsTrivial and IsNonTrivial;
+        if Length(gens)<=1 then
+          typ:=typ and IsCyclic;
+          typename:="defaultNontrivialCyclicGroupType";
+        else
+          typename:="defaultFinitelyGeneratedNontrivialGroupType";
+        fi;
+      else
+        typ:=typ and IsTrivial and HasIsNonTrivial;
+        typename:="defaultTrivialGroupType";
       fi;
-      typ:=fam!.defaultGroupType;
+    elif Length(gens)<=1 then
+      typ:=typ and IsCyclic;
+      typename:="defaultCyclicGroupType";
     fi;
+  fi;
+  if not IsBound(fam!.(typename)) then
+    typ:=NewType(fam,typ);
+    fam!.(typename):=typ;
+  else
+    typ:=fam!.(typename);
+  fi;
 
-    G:=rec();
-    ObjectifyWithAttributes(G,typ,GeneratorsOfMagmaWithInverses,AsList(gens));
+  G:=rec();
+  ObjectifyWithAttributes(G,typ,GeneratorsOfMagmaWithInverses,AsList(gens));
 
-    return G;
-    end );
+  return G;
+end );
 
 InstallMethod( GroupWithGenerators,
     "generic method for collection and identity element",
     IsCollsElms, [ IsCollection, IsMultiplicativeElementWithInverse ],
-    function( gens, id )
-    local G,fam,typ;
-
-    fam:=FamilyObj(gens);
-    if IsFinite(gens) then
-      if not IsBound(fam!.defaultFinitelyGeneratedGroupWithOneType) then
-        fam!.defaultFinitelyGeneratedGroupWithOneType:=
-          NewType(fam,IsGroup and IsAttributeStoringRep
-                      and HasGeneratorsOfMagmaWithInverses
-                      and IsFinitelyGeneratedGroup and HasOne);
-      fi;
-      typ:=fam!.defaultFinitelyGeneratedGroupWithOneType;
-    else
-      if not IsBound(fam!.defaultGroupWithOneType) then
-        fam!.defaultGroupWithOneType:=
-          NewType(fam,IsGroup and IsAttributeStoringRep
-                      and HasGeneratorsOfMagmaWithInverses and HasOne);
+function( gens, id )
+local G,fam,typ,typename;
+
+  fam:=FamilyObj(gens);
+
+  typ:=IsGroup and IsAttributeStoringRep 
+        and HasIsEmpty and HasGeneratorsOfMagmaWithInverses and HasOne;
+  typename:="defaultGroupWithOneType";
+
+  if IsFinite(gens) then
+    typ:=typ and IsFinitelyGeneratedGroup;
+    typename:="defaultFinitelyGeneratedGroupWithOneType";
+
+    if Length(gens)>0 and CanEasilyCompareElements(gens) then
```

Comment:
Use `>=0` to deal with empty collections? Or even better, as this also works if `CanEasilyCompareElements(gens)` is false, start like this:
```gap
if Length(gens) = 0 then
 ...
elif CanEasilyCompareElements(gens) then
 ...
elif Length(gens) = 1 then
 ...
fi;
```

Reply:
But there is a specialized method below for the case of `IsEmpty` for the collection `gens`. Wouldn't it override it?

Reply:
That method is only for empty lists, but not for empty collections-that-aren't-lists. Though these are rather rare and special beast.

But as I said before, it would also be nice if all this logic would not have to be repeated in almost identical form several times, but rather was put into one (or at least very few) helper functions which can be called from multiple places.

Reply:
The deductions logic is now in a separate function.